### PR TITLE
Properly generate secrets files

### PR DESCRIPTION
**Pull request by cloader89**
This allows for the secrets file to be merged with the existing
secrets files without leaving any of the new secrets ungenerated.
This cleans up from our old way of copying the existing secrets
file from the previous installation. 

Connects http://github.com/rcbops/u-suk-dev#475

### CODE REVIEW

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/staging.horse/build-071e5a3884/docs/private-cloud/release-notes/)

**Comment by cloader89** [comment]
Still doing some testing, but seemed to work on my AIO.

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/staging.horse/build-7c4341df9b/docs/private-cloud/release-notes/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-3ada561aa2/docs/private-cloud/release-notes/).

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/staging.horse/build-fb83520c58/docs/private-cloud/release-notes/)

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/staging.horse/build-c1ee8a90eb/docs/private-cloud/release-notes/)

**Comment by prometheanfire** [comment]
This looks like it should be fine.  Though could you also run the following instead?

`migrate-yaml.py --defaults $RPCD_DIR/etc/openstack_deploy/user_rpco_secrets.yml --overrides /etc/openstack_deploy/user_extras_secrets.yml --output-file /etc/openstack_deploy/user_rpco_secrets.yml`

**Comment by cloader89** [comment]
@prometheanfire  That's what I originally tried. It did not carry over the new vars from `user_rpco_secrets.yml ` defaults

**Comment by prometheanfire** [comment]
ok, perhaps the migrate script needs fixing then

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/staging.horse/build-d1114e2cc3/docs/private-cloud/release-notes/)

**Comment by cloader89** [comment]
recheck_all

**Comment by cloader89** [comment]
recheck_all

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-dc2c008bb7/docs/private-cloud/release-notes/).

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/staging.horse/build-0a14293e8a/docs/private-cloud/release-notes/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-694142e669/docs/private-cloud/release-notes/).

**Comment by alextricity25** [comment]
Doc Issue is here: https://github.com/rackerlabs/docs-rpc/issues/764

File: releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+---
+features:
+  - The update-yaml.py script now contains an optional
+    argument, --output-file, for specifying an output
+    file.
```
Comment (major) [inline releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml]: Small nit. It might be nice to get `update-yaml.py` and `--output-file` monospaces so they're easier to see.

File: releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+upgrade:
+  - Any new secret variables that may be introduced in
+    later releases is now handled by the update-secrets.sh
+    script. test-upgrade.sh and deploy.sh have been updated
```
Comment (major) [inline releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml]: Nit: Wrap double backticks around `test-upgrade.sh` so that it is monospaced.

File: scripts/update-secrets.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# If the user_rpco_secrets.yml file exists in the user config, then merge that with
+# the user_rpco_secrets.yml in tree. Otherwise, see if the user_extras_secrets.yml
+# file exists and if so, use that.
+if [[ -f "/etc/openstack_deploy/user_rpco_secrets.yml" ]]; then
+  OVERRIDES="/etc/openstack_deploy/user_rpco_secrets.yml"
+elif [[ -f "/etc/openstack_deploy/user_extras_secrets.yml" ]];then
+  OVERRIDES="/etc/openstack_deploy/user_extras_secrets.yml"
+else
+  echo "Cannot find user defined rpco secrets file! Please copy an rpco secrets file to /etc/openstack_deploy/"
+fi
```
Comment (major) [inline scripts/update-secrets.sh]: Should the script exit here since the files aren't present?

Comment (cloader89) [inline scripts/update-secrets.sh]: Yes will add that. Thanks.

File: releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml (hunk 2 above)
Comment (cloader89) [inline releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml]: Ok, thanks.

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/staging.horse/build-90286ea25b/docs/private-cloud/release-notes/)

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/staging.horse/build-3b7d321ada/docs/private-cloud/release-notes/)

File: scripts/update-yaml.py [hunk 4]
```diff
@@ -25,6 +25,8 @@ def parse_args():
                              'configuration.')
     parser.add_argument('overrides',
                         help='The path to the yaml file with overrides.')
+    parser.add_argument('--output-file',
+                        help='The output file')
```
Comment (sigmavirus24) [inline scripts/update-yaml.py]: This help could be more descriptive.

File: scripts/update-yaml.py [hunk 5]
```diff
@@ -45,12 +47,16 @@ def get_config(path):
     else:
         return data
 
-
 if __name__ == '__main__':
```
Comment (sigmavirus24) [inline scripts/update-yaml.py]: Why did you remove the newline above this?

Comment (cloader89) [inline scripts/update-yaml.py]: I originally was adding a new method, but I removed it by accident. Fixing.

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/staging.horse/build-e214dfcb90/docs/private-cloud/release-notes/)

**Comment by cloader89** [comment]
recheck_upgrade

**Comment by cloader89** [comment]
tempest.lib.exceptions.ServerFault: Got server fault recheck_ceph

**Comment by major** [comment]
recheck_upgrade

Looks like there was a failure with ssh during the instance spawn.

**Comment by cloader89** [comment]
NOTIFIED: [horizon_extensions | Restart apache2] ****************************** 
17:52:16 changed: [jrpcaioiad-b9b_horizon_container-98589c0e]
17:52:35 failed: [jrpcaioiad-b9b_horizon_container-adf057de] => {"failed": true}
17:52:35 msg: (98)Address already in use: AH00072: make_sock: could not bind to address [::]:80
17:52:35 (98)Address already in use: AH00072: make_sock: could not bind to address 0.0.0.0:80
17:52:35 no listening sockets available, shutting down

recheck_upgrade

**Comment by sigmavirus24** [comment]
This looks (:eyes:-only) good but I haven't given it a test 🚗 . Preliminary 👍 from me

**Comment by d34dh0r53** [comment]
Needs to be 🔨 but after that I'm 😁 and give it my 👍

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/staging.horse/build-3af7bf9941/docs/private-cloud/release-notes/)

**Comment by cloader89** [comment]
recheck_upgrade -- maas API

**Comment by cloader89** [comment]
recheck_swift

**Comment by cloader89** [comment]
recheck_swift

**Comment by cloader89** [comment]
recheck_ceph

File: releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+---
+features:
+  - The ``update-yaml.py`` script now contains an optional
+    argument, ``--output-file``, for specifying an output
+    file.
+upgrade:
+  - Any new secret variables that may be introduced in
+    later releases is now handled by the ``update-secrets.sh``
```
Comment (prometheanfire) [inline releasenotes/notes/handle-merging-new-secrets-0e1fd06d86f0abbd.yaml]: s/is/are

File: scripts/test-upgrade.sh [hunk 7]
```diff
@@ -50,8 +50,7 @@ fi
 #       liberty install. This won't cause any problems since there aren't any new
 #       secret variables in mitaka that are not defined in liberty.
 if [[ -f "/etc/openstack_deploy/user_extras_secrets.yml" ]]; then
```
Comment (prometheanfire) [inline scripts/test-upgrade.sh]: I'm not sure if this if statement is needed anymore.

File: scripts/update-secrets.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+if [[ -f "/etc/openstack_deploy/user_rpco_secrets.yml" ]]; then
+  OVERRIDES="/etc/openstack_deploy/user_rpco_secrets.yml"
+elif [[ -f "/etc/openstack_deploy/user_extras_secrets.yml" ]]; then
+  OVERRIDES="/etc/openstack_deploy/user_extras_secrets.yml"
+else
+  echo "Cannot find user defined rpco secrets file! Please copy an rpco secrets file to /etc/openstack_deploy/"
+  exit
+fi
+
+python /opt/rpc-openstack/scripts/update-yaml.py --output-file /etc/openstack_deploy/user_rpco_secrets.yml $RPCD_DIR/etc/openstack_deploy/user_rpco_secrets.yml $OVERRIDES
```
Comment (prometheanfire) [inline scripts/update-secrets.sh]: would prefer if it called python2 or python2.7 for future proofing.

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-7abb6deebf/docs/private-cloud/release-notes/).

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/staging.horse/build-65d2b03405/docs/private-cloud/release-notes/)

**Comment by tomc603** [comment]
👎  bashate is failing complaining about trailing whitespace.

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-6091b75d28/docs/private-cloud/release-notes/).

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-7457139870/docs/private-cloud/release-notes/).

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/staging.horse/build-9402800dc7/docs/private-cloud/release-notes/).

**Comment by cloader89** [comment]
recheck_ceph

**Comment by cloader89** [comment]
ceph.ceph-osd was NOT installed successfully. recheck_ceph

**Comment by cloader89** [comment]
ready to merge/review @tomc603

**Comment by tomc603** [comment]
Great! 👍
